### PR TITLE
[RFC] test: fnamemodify() continued

### PR DESCRIPTION
**Pull request by janlazo**
Continue #7829 but I will work on the fix to make the tests pass.

### CODE REVIEW

**Comment by janlazo** [comment]
```
[  FAILED  ] 1 test, listed below:
[  FAILED  ] ...rojects/neovim/test/functional\eval\fnamemodify_spec.lua @ 11: fnamemodify() works
...rojects/neovim/test/functional\eval\fnamemodify_spec.lua:18: Expected objects to be the same.
Passed in:
(string) 'C:/projects/neovim'
Expected:
(string) 'C:/'
stack traceback:
	...rojects/neovim/test/functional\eval\fnamemodify_spec.lua:18: in function <...rojects/neovim/test/functional\eval\fnamemodify_spec.lua:11>
```

**Comment by janlazo** [comment]
Is there path expansion for `/`? I don't know why `cwd` is appended.

**Comment by ZyX-I** [comment]
I am not sure whether `/` *should* mean `C:` at all, where this idea is from, and what if there is no `C:`? I also know that Windows does have paths like `\\remote\dir`, even if there is no actual `\\` (not sure, did not check) what `fnamemodify()` does to `\\remote\dir` is worth testing.

**Comment by janlazo** [comment]
I assume `/` is relative to the current drive. Current directory of the test suite is on 'C:/projects/neovim' (perhaps get it via `getcwd()`?) so `/` should point to `C:`.
Not sure about [UNC paths](https://en.wikipedia.org/wiki/Path_(computing)#Uniform_Naming_Convention) on Windows. Is that tested on Appveyor?

**Comment by janlazo** [comment]
Why did that work (on Appveyor)? `PATHSEP`  changed because of `shellslash`?

**Comment by janlazo** [comment]
The test for tui focus gained/lost is flaky. Can that run in flaky builds only?

**Comment by janlazo** [comment]
Should be `:8` supported? I can try porting my [workaround](https://github.com/junegunn/fzf.vim/commit/7e92b73d03cd008e248ad261208475b45bc2f128#diff-7274b7c6c87526542ef9087f2e997dfaR40) from fzf.vim.

**Comment by janlazo** [comment]
Naive implementation is as follows:
1. Create a new variable, `use_shortname`, initialize to 0
2. if `:8` is found, set `use_shortname` to 1 and then continue as before.
3. After going through all the flags and before returning, if `use_shortname == 1` and if compiled on Windows,, convert filepath to dos shortname using my workaround.

`shellescape` is enough for now but I need https://github.com/neovim/neovim/pull/7786/commits/ca5f2d73be570666257dcc666164be19fffff45d for literal escaping.

File: test/functional/legacy/fnamemodify_spec.lua [hunk 1]
```diff
@@ -16,17 +14,27 @@ describe('filename modifiers', function()
 
     source([=[
       func Test_fnamemodify()
-        let tmpdir = resolve('/tmp')
+        if has('win32')
+          set shellslash
+          let tmpdir = expand('$LOCALAPPDATA/Temp')
+          call assert_equal(tmpdir, expand('$USERPROFILE/AppData/Local/Temp'))
+        else
+          let tmpdir = resolve('/tmp')
```
Comment (justinmk) [inline test/functional/legacy/fnamemodify_spec.lua]: $TMPDIR is passed to the tests, can it be used instead? It is local to the project directory. (`./Xtest-tmpdir`)

**Comment by justinmk** [comment]
LGMT, is this RDY? 

If `:8` needs core changes let's do it in another PR.

File: test/functional/legacy/fnamemodify_spec.lua (hunk 1 above)
Comment (janlazo) [reply test/functional/legacy/fnamemodify_spec.lua]: Yes.

**Comment by janlazo** [comment]
`:8` needs core changes so I'm okay with a separate PR. I'm not sure which implementation to go with after looking at Vim's implementation.

I'll update the legacy test to use `$TMPDIR` then we're RDY.

**Comment by janlazo** [comment]
`TMPDIR` is not set for QuickBuild?

**Comment by justinmk** [comment]
TMPDIR is set to a non-local path for quickbuild (don't remember why).  `helpers.tmpdir_is_local()` could be useful.

File: test/functional/eval/fnamemodify_spec.lua [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  before_each(clear)
+
+  teardown(function()
+    os.remove('Xtest-fnamemodify.txt')
+  end)
+
+  it('works', function()
+    if iswin() then
+      local drive_f = io.popen('for %P in (%CD%) do @echo %~dP', 'r')
+      local drive = string.gsub(drive_f:read('*a'), '[\n\r]', '')
```
Comment (justinmk) [inline test/functional/eval/fnamemodify_spec.lua]: on master ( https://github.com/neovim/neovim/commit/499c9a15531b7a0e9736a395e8d401ceab3d24d2 ) there is now `helpers.pathroot()`, can that be used instead?

**Comment by justinmk** [comment]
@janlazo besides the note about `helpers.pathroot()`, today I pushed some changes related to $TMPDIR. Could you rebase this to make sure I didn't cause a problem?

File: test/functional/eval/fnamemodify_spec.lua (hunk 2 above)
Comment (janlazo) [reply test/functional/eval/fnamemodify_spec.lua]: Yes. I can strip out the pathsep and append a forward slash for `shellslash`.

**Comment by justinmk** [comment]
quickbuild failure looks unrelated (the quickbuild env was upgraded recently and we're working through it).

So I'll wait for the `helpers.pathroot` change then merge this.

File: test/functional/eval/fnamemodify_spec.lua [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      root = string.sub(root, 1, -2)..'/'
+      eq(root, fnamemodify([[\]], ':p:h'))
+      eq(root, fnamemodify([[\]], ':p'))
+      eq(root, fnamemodify([[/]], ':p:h'))
+      eq(root, fnamemodify([[/]], ':p'))
+    end
+  end)
+
+  it(':8 works', function()
+    eq('Xtest-fnamemodify.txt', fnamemodify([[Xtest-fnamemodify.txt]], ':8'))
```
Comment (justinmk) [inline test/functional/eval/fnamemodify_spec.lua]: This doesn't test much...

Comment (janlazo) [reply test/functional/eval/fnamemodify_spec.lua]: The expected result there should be the shortened DOS name but I haven't started on `:8`.

**Pull request merged**